### PR TITLE
[tt-train] Support MFU metric and more models

### PR DESCRIPTION
**Pull request by kevinwuTT**
### Summary

* Add character-level version of nanollama3
* Dynamically extract the amount of available device memory to use as a metric
* Parse MFU metric from step output
* Dynamically extract the card type used in machine
* Add option to run `dataset_to_tokens.py` script before running the model

Changes to `run_models.py` script used for debugging and experiments:
* Generate Github summary of run times
* Add parameters to filter or exclude tests
* Add parameter to insert extra args to existing tests

New model list:
Screenshot 2026-05-01 at 4 21 17 PM

### Notes for reviewers

* Since the default training configurations for most of the relevant models have changed and resulted in increased memory usage, most of the workload is now being done on P150. 
  * The training script allows overwriting certain options. For example, decreasing batch_size could make the model fit for the other cards. Do we want to do that?
* In order to maintain a reasonable runtime length below 4 hours, Qwen 1.7B and TinyDeepSeek has their max steps decreased. Is this fine? 

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-kevinwu/model_run_mfu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ctr-kevinwu/model_run_mfu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ctr-kevinwu/model_run_mfu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-kevinwu/model_run_mfu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ctr-kevinwu/model_run_mfu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ctr-kevinwu/model_run_mfu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ctr-kevinwu/model_run_mfu)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ctr-kevinwu/model_run_mfu)

### CODE REVIEW

File: tt-train/scripts/run_models.py [hunk 1]
```diff
@@ -143,17 +186,55 @@ def main() -> int:
...
+            if ret_code != 0:
+                failing_models.append(str(log_path))
+                print(f"Subprocess dataset_to_tokens.py failed. Return code {ret_code}")
+                continue
+
+        cmd = process_binary_path(binary, tt_metal_runtime_root) + args + extra_args
+        print("=" * 100)
+        print(f"Running {model_filename}")
+        print("-" * 100)
+        print(f"cmd: {' '.join(cmd)}")
```
[the remark this answers is not in the document]
Comment (kevinwuTT) [reply tt-train/scripts/run_models.py:241]: This is fine. `extra_args` is used for debugging purposes.

File: tt-train/scripts/run_models_config.yaml [hunk 2]
```diff
@@ -30,10 +30,68 @@ models:
...
+      - "-c ${TT_METAL_RUNTIME_ROOT}/tt-train/configs/training_configs/training_shakespeare_qwen3_17b.yaml"
+      - "--track_memory"
+    skip-arch:
+      - "n150"
+      - "n300"
+      - "p100"
+    tokenize:
+      text_file: "data/shakespeare.txt"
+      hf_tokenizer: "Qwen/Qwen3-1.7B"
+      output_file: "data/tokenized_shakespeare_qwen3"
```
[the remark this answers is not in the document]
Comment (kevinwuTT) [reply tt-train/scripts/run_models_config.yaml:70]: This should be fine since. Each run will sequentially generate the tokenized file and use it immediately.

**Review by roseli-TT — APPROVED** [review]

**Review by jyuanTT — APPROVED** [review]

File: tt-train/configs/model_configs/nanollama3_char.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+transformer_config:
```
Comment (jmalone-tt) [inline tt-train/configs/model_configs/nanollama3_char.yaml:1]: The only difference with nanollama3.yaml I see is omitting the vocab size: 32000 - do we need both model configs?

File: tt-train/configs/training_configs/training_shakespeare_nanollama3_char.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+training_config:
+  project_name: "tt_train_nano_gpt" # not really nanogpt, but want to use the same wandb project name for now
```
Comment (jmalone-tt) [inline tt-train/configs/training_configs/training_shakespeare_nanollama3_char.yaml]: I think we're actually fixing this now - let's do something that's accurate instead

File: tt-train/scripts/analyze_memory.py [hunk 5]
```diff
@@ -45,6 +45,18 @@ def extract_number_of_parameters(content: str, start_pos: int) -> Optional[int]:
     return None
 
 
+def extract_available_device_memory_mb(content: str, start_pos: int) -> Optional[int]:
+    """Extract the available device memory from logs before the memory summary"""
```
Comment (jmalone-tt) [inline tt-train/scripts/analyze_memory.py:49]: I'm assuming we are parsing from logs and not using a device API because this is running after the fact - is that right?

File: tt-train/scripts/run_models.py [hunk 6]
```diff
@@ -46,11 +48,24 @@ def get_git_commit_hash() -> str:
     return sha
 
 
-def run_and_save_log(args: list[str], log_path: Path):
+def get_card_type() -> str:
+    """Get the card type by reading from the sysfs or from CI_RUNNER_LABEL. Return exception if fails."""
```
Comment (jmalone-tt) [inline tt-train/scripts/run_models.py]: It seems like we have some existing APIs for this - can probably use one of them instead of reimplementing
https://deepwiki.com/search/whats-the-api-for-figuring-out_66c9785c-3df0-4a56-b27b-62b5d22330e1?mode=fast

File: tt-train/scripts/tt_train_metrics.py [hunk 7]
```diff
@@ -110,6 +110,10 @@ class TtTrainMetricsData(BaseModel):
         default=None,
         description="Iteration time at the 99th percentile (ms).",
     )
+    mfu: Optional[float] = Field(
+        default=None,
+        description="Model flops utilization.",
```
Comment (jmalone-tt) [inline tt-train/scripts/tt_train_metrics.py]: nit: FLOPs

File: tt-train/configs/model_configs/nanollama3_char.yaml (hunk 3 above)
Comment (kevinwuTT) [reply tt-train/configs/model_configs/nanollama3_char.yaml:1]: This [PR](https://github.com/tenstorrent/tt-metal/pull/42654) made the separation between BPE and Char tokenization more explicit. BPE tokenization needs to have the `vocab_size` be `> 0`. Char tokenization needs to omit `vocab_size` or set it to 0. `nanollama3.yaml` has `vocab_size: 32000` defined, so it is meant to be used with BPE tokenization. BPE tokenization requires using the `tt-train/tools/dataset_to_tokens.py` script to generate a dataset with a pre-trained checkpoint before running the training script. Previously, that was inconsistent, and I could run the training script with `nanollama3.yaml` without running the pretokenization script first. So to remain consistent with previous runs, I copied `nanollama3.yaml` and created the "Char" version of it. 

If `nanollama3.yaml` is meant to be used with char tokenization, then it will be easier to remove `vocab_size` from that instead of creating new config. Otherwise, I'm not aware of a NanoLlama pre-trained checkpoint to use with the `tt-train/tools/dataset_to_tokens.py` script if `nanollama3.yaml` is meant to be used with BPE tokenization.

File: tt-train/scripts/analyze_memory.py (hunk 5 above)
Comment (kevinwuTT) [reply tt-train/scripts/analyze_memory.py:49]: `analyze_memory.py` would be parsing from logs yeah. But I added a function that uses device API in `tt-train/sources/ttml/ttml/common/utils.py`. `train_nanogpt.py` would call this function and append to the log. Something like this:

```
def get_available_device_memory_in_bytes() -> int:
    """Get the total amount of device DRAM available on the system."""
    device = ttml.autograd.AutoContext.get_instance().get_device()
    dram_view = ttnn.device.get_memory_view(device, ttnn.BufferType.DRAM)
    total_dram = dram_view.total_bytes_per_bank * dram_view.num_banks * ttnn.get_num_devices()
    return total_dram
```

File: tt-train/configs/training_configs/training_shakespeare_nanollama3_char.yaml (hunk 4 above)
Comment (kevinwuTT) [reply tt-train/configs/training_configs/training_shakespeare_nanollama3_char.yaml]: I changed it to `tt_train_nano_llama`. WDYT?

File: tt-train/scripts/run_models.py (hunk 6 above)
Comment (kevinwuTT) [reply tt-train/scripts/run_models.py]: `from model_tracer.generic_ops_tracer import get_machine_info` seems to work locally. Will test it on CI as well to make sure it's importable.

Comment (kevinwuTT) [reply tt-train/scripts/run_models.py]: Looks like I do have to point `PYTHONPATH` to root of the repo if I install ttnn via wheel, which is what CI is doing. If that's fine, then I'll keep this change.

**Review by philei-tt — APPROVED** [review]
Thanks! Let's discuss adding "galaxy-sized" models
